File: 2Silver/2122/3FebruaryC/problem2.py
"""
ID: 06jajun
LANG: PYTHON3
TASK: meetings
"""
from collections import defaultdict


def main():
    n = int(input())
    px, py = [int(c) for c in input().split()]
    instructions = []
    m = n // 2
    all = defaultdict(lambda: [])
    res = [0] * n
    for i in range(n):
        instructions.append([int(c) for c in input().split()])

    def recursea(i=0, x=0, y=0, k=0):
        if i == m:
            all[(px - x, py - y)].append(k)
        else:
            nx, ny = instructions[i]
            recursea(i + 1, x + nx, y + ny, k + 1)
            recursea(i + 1, x, y, k)
    recursea()
    # print(all)

    def recurseb(i=m, x=0, y=0, k=-1):
        if i == n:
            for j in all[(x, y)]:
                res[j + k] += 1
        else:
            nx, ny = instructions[i]
            recurseb(i + 1, x + nx, y + ny, k + 1)
            recurseb(i + 1, x, y, k)
    recurseb()
    print('\n'.join([str(c) for c in res]))


if __name__ == '__main__':
    main()

File: 2Silver/2122/3FebruaryC/problem2.cpp
#include <iostream>
#include <unordered_map>
#include <tuple>
#include <array>
#include <vector>
using namespace std;

int n, px, py;
int m;
unordered_map<long, vector<int>> all;
vector<int> res;

void recursea(int i, int x, int y, int k, int instructions[40][2]) {
  if (i == m) {
    long check = (px - x) * 1000000000 + py - y;
    all[check].push_back(k);
  }
  else {
    int nx = instructions[i][0];
    int ny = instructions[i][1];
    recursea(i + 1, x + nx, y + ny, k + 1, instructions);
    recursea(i + 1, x, y, k, instructions);
  }
}

void recurseb(int i, int x, int y, int k, int instructions[40][2]) {
  if (i == n) {
    long check = 1000000000 * x + y;
    for (auto j: all[check]) res[j + k] += 1;
  }
  else {
    int nx = instructions[i][0];
    int ny = instructions[i][1];
    recurseb(i + 1, x + nx, y + ny, k + 1, instructions);
    recurseb(i + 1, x, y, k, instructions);
  }
}

int main() {
  cin >> n >> px >> py;
  m = n / 2;
  int instructions[40][2];
  for (int i=0; i<n; i++) {
    cin >> instructions[i][0] >> instructions[i][1];
  }
  for (int i=0; i<n; i++) {
    res.push_back(0);
  }

  recursea(0, 0, 0, 0, instructions);

  recurseb(m, 0, 0, -1, instructions);
  for (auto x: res) {
    cout << x << endl;
  }
}
